hal/src/dma/channel.rs

```diff
@@ -0,0 +1,584 @@
+
+        Ok(())
+    }
+
+    /// Set DMA priority
+    pub fn set_priority(&mut self, priority: Priority) {
+        // TODO(tarcieri): confirm it's ok to acquire/hold a lock here
+        // The original Adafruit code did not
+        let dmac = self.controller.lock();
+        dmac.set_priority(self.channel_id, priority);
+    }
+
+    /// Deallocate DMA channel
+    // TODO(adafruit): should this delete/deallocate the descriptor list?
+    pub fn free(&mut self, clock_source: &mut ClockSource, nvic: &mut NVIC) -> Result<(), Error> {
+        // job_status is volatile
+        let dmac = self.controller.lock();
+
+        // Can't leave when busy
+        if self.job_status.get() == Err(Error::Busy) {
+            return Err(Error::Busy);
+        }
+
+        dmac.release_channel(self.channel_id, clock_source, nvic)
+    }
+
+    /// Start DMA transfer job.
+    pub fn start_job(&mut self) -> Result<(), Error> {
+        // job_status is volatile
+        let dmac = self.controller.lock();
+
+        if self.job_status.get() == Err(Error::Busy) {
+            // Resource is busy
+            return Err(Error::Busy);
+        }
+
+        if !self.has_descriptors
+            || self.descriptor_list.descriptor(self.channel_id).btcnt.get() <= 0
+        {
+            return Err(Error::InvalidArg); // Bad transfer size
+        }
+
+        let mut interrupt_mask = 0;
+
+        if self.callbacks.transfer_error.is_some() {
+            interrupt_mask |= 1;
+        }
+
+        if self.callbacks.transfer_done.is_some() {
+            interrupt_mask |= 1 << 1;
+        }
+
+        if self.callbacks.channel_suspend.is_some() {
+            interrupt_mask |= 1 << 2;
+        }
+
+        self.job_status.set(Err(Error::Busy));
+        dmac.set_interrupt_mask(self.channel_id, interrupt_mask);
+
+        Ok(())
+    }
+
+    /// Set and enable callback function for DMA object. This can be called
+    /// before or after channel and/or descriptors are allocated, but needs
+    /// to be called before job is started.
+    pub fn set_callback(&mut self, cb_type: CallbackType, cb: fn(&mut Channel)) {
+        match cb_type {
+            CallbackType::TransferError => self.callbacks.transfer_error = Some(cb),
+            CallbackType::TransferDone => self.callbacks.transfer_done = Some(cb),
+            CallbackType::ChannelSuspend => self.callbacks.channel_suspend = Some(cb),
+        }
+    }
+
+    /// Abort
+    pub fn abort(&mut self) {
+        let dmac = self.controller.lock();
+        dmac.abort(self.channel_id);
+        self.job_status.set(Err(Error::Aborted));
+    }
+
+    /// Set DMA peripheral trigger.
+    /// This can be done before or after channel is allocated.
+    pub fn set_trigger(&mut self, trigger: TRIGSRCW) {
+        let dmac = self.controller.lock();
+
+        // Save value for `allocate()`
+        self.peripheral_trigger = match self.peripheral_trigger {
```

Comment:
```suggestion
        self.peripheral_trigger = match trigger {
```

also on `master`, in the newly generated pac `TRIGSRCW` -> `TRIGSRC_A` and has copy

Reply:
@Gentoli thanks. this PR is super out of date now though! Any interest in helping pick it back up?